algebird-core/src/main/scala/com/twitter/algebird/TDigest.scala

```diff
@@ -0,0 +1,275 @@
+   * @param x The numeric data value to include in the sketch
+   * @return the updated sketch
+   */
+  def +[N](x: N)(implicit num: Numeric[N]): TDigest = this.+((x, 1))
+
+  /**
+   * Returns a new t-digest with new pair (x, w) included in its sketch.
+   * @param xw A pair (x, w) where x is the numeric value and w is its weight
+   * @return the updated sketch
+   * @note This implements 'algorithm 1' from:
+   * Computing Extremely Accurate Quantiles Using t-Digests,
+   * Ted Dunning and Otmar Ertl,
+   * https://github.com/tdunning/t-digest/blob/master/docs/t-digest-paper/histo.pdf
+   */
+  def +[N1, N2](xw: (N1, N2))(implicit num1: Numeric[N1], num2: Numeric[N2]): TDigest = {
+    val s = this.update(xw)
+    if (s.nclusters <= R) s
+    else {
+      // too many clusters: attempt to compress it by re-clustering
+      val ds = TDigest.shuffle(s.clusters.toVector)
+      ds.foldLeft(TDigest.empty(delta))((d, e) => d.update(e))
+    }
+  }
+
+  /**
+   * Add this digest to another
+   * @param that The right-hand t-digest operand
+   * @return the sum of left and right digests, as defined by TDigestSemigroup
+   * @see TDigestSemigroup
+   */
+  def ++(that: TDigest): TDigest = TDigestSemigroup.plus(this, that, this.delta)
+
+  // This is most of 'algorithm 1', except for re-clustering which is factored out to avoid
+  // recursive calls during a reclustering phase
+  private def update[N1, N2](xw: (N1, N2))(implicit num1: Numeric[N1], num2: Numeric[N2]) = {
+    val xn = num1.toDouble(xw._1)
+    var wn = num2.toDouble(xw._2)
+    require(wn > 0.0, "data weight must be > 0")
+
+    // Get the current cluster nearest to incoming (xn)
+    // Note: 'near' will have length 0,1, or 2:
+    // length 0 => current cluster map was empty (no data yet)
+    // length 1 => exactly one cluster was closest to (xn)
+    // length 2 => (xn) was mid-point between two clusters (both are returned, in key order)
+    val near = clusters.nearest(xn)
+
+    if (near.isEmpty) {
+      // our map is empty, so insert this pair as the first cluster
+      TDigest(delta, nclusters + 1, clusters + ((xn, wn)))
+    } else {
+      // compute upper bounds for cluster masses, from their quantile estimates
+      var massPS = clusters.prefixSum(near.head._1, open = true)
+      val massTotal = clusters.sum
+      val s = near.map {
+        case (c, m) =>
+          val q = (massPS + m / 2.0) / massTotal
+          val ub = massTotal * delta * q * (1.0 - q)
+          massPS += m
+          Cluster(c, m, ub)
+      }
+
+      // assign new mass (wn) among the clusters
+      var cmNew = Vector.empty[(Double, Double)]
+      TDigest.shuffle(s).foreach { clust =>
+        if (wn <= 0.0) {
+          // if we have already distributed all the mass, remaining clusters unchanged
+          cmNew = cmNew :+ ((clust.centroid, clust.mass))
+        } else if (xn == clust.centroid) {
+          // if xn lies exactly on the centroid, add all mass in regardless of bound
```

Comment:
What's the justification for this?  It doesn't seem to come from the paper.

In fact, there is explicit acknowledgment that two clusters could have the same centroid:
> For centroids with identical means, order of creation is used as a tie-breaker to allow an unambiguous ordering